QoL: Buildtype variables for looping over buildtypes

Neumann-A opened the pull request:
While working on the qt scripts I often defined those variables below to avoid unnecessary code duplication to loop over the different buildtypes

TODO:
- [ ] Rename `VCPKG_BUILD_TRIPLET_*` to something different ?
- [ ] Remove `VCPKG_BUILDTREE_TRIPLET_DIR_` ?

Neumann-A committed:
define buildtype variables for easier looping over them

Neumann-A committed:
Merge branch 'master' into build_variables

Neumann-A committed:
Merge branch 'master' into build_variables

ras0219-msft commented:
Thanks for the PR!

This _in principle_ seems like it could be a useful abstraction, but I'd like to see it demonstrated in multiple contexts before merging.

Specifically, if it turns out to only be useful in Qt5, then it should be private to Qt5. However, if it can be used to make code significantly cleaner and clearer in more contexts, I'd like to see this PR introduce at least two of those refactorings.

Neumann-A commented:
TODO (only 2 of N):
- [ ] Refactor vcpkg_build_cmake
- [ ] Refactor vcpkg_build_msbuild
- [ ] Refactor vcpkg_build_qmake
- [ ] Refactor vcpkg_configure_cmake
- [ ] Refactor vcpkg_configure_meson
- [ ] Refactor vcpkg_configure_qmake
- [ ] Refactor vcpkg_fixup_cmake_targets
- [ ] Refactor vcpkg_install_meson (a bit short for a refactor, but consistency is important)
- [ ] Refactor vcpkg_install_msbuild
- [ ] Refactor an unknown number of ports which might benift from the PR (purely optional; but at least qt if the update to 5.12.4 PR is merged.)

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

# Conflicts:
#	scripts/cmake/vcpkg_common_definitions.cmake

Neumann-A committed:
refactored vcpkg_configure_cmake with new variables

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

Neumann-A committed:
add/update documentation

Neumann-A committed:
refactored vcpkg_configure_qmake

Neumann-A committed:
refactored qt-5base configure_qt

Neumann-A committed:
refactored install_qt in qt5-base

Neumann-A committed:
refactor vcpkg_build_cmake

Neumann-A committed:
note about the possibility of introducing an option mapping

Neumann-A committed:
refactored vcpkg_install_msbuild

Neumann-A committed:
fix linux regressions

Neumann-A committed:
bump qt control

JackBoosY commented:
Please resolve these conflicts.

Thanks.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

# Conflicts:
#	ports/qt5-base/CONTROL
#	ports/qt5-base/cmake/configure_qt.cmake
#	ports/qt5-base/cmake/install_qt.cmake
#	scripts/cmake/vcpkg_build_cmake.cmake
#	scripts/cmake/vcpkg_configure_cmake.cmake
#	scripts/cmake/vcpkg_configure_qmake.cmake

Neumann-A committed:
use globaly defined variables in qmake configure

JackBoosY commented:
Pinging @Neumann-A  for response. Is work still being done for this PR?

Neumann-A commented:
@JackBoosY: You put the PR back into Draft?!?! I consider it finished for the time unless there is something which needs to ne fixed which I don't know of or you want some changes to it. Maybe run a full rebuild of everything to check if everything is working as intended.

JackBoosY commented:
@Neumann-A Sorry, my fault.

JackBoosY commented:
/azp run

JackBoosY commented:
Could you select some ports to test these functions?

Neumann-A committed:
[fftw3] remove deprecated function for cmake CI test

Neumann-A committed:
[3fd] remove deprecated function for CI rebuild (msbuild)

Neumann-A committed:
[qcustomplot] remove deprecated include for CI rerun (qmake)

Neumann-A committed:
[vcpkg/scripts] add missing ) in vcpkg_configure_qmake

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

Neumann-A committed:
[vcpkg/scripts] fix cmake configure

Neumann-A committed:
ws change to rerun ci

Neumann-A committed:
[fftw3] bump control.

Neumann-A committed:
[vcpkg/scripts] small change to vcpkg_fixup_cmake_targets to trigger a full rebuild

Neumann-A committed:
[vcpkg/scripts/cmake] use same quoting as before.

Neumann-A committed:
[vcpkg/scripts/cmake] change to fixup to trigger full CI rebuild

Neumann-A committed:
[vcpkg/scripts/msbuild] fix "EXES" "AND" error

Neumann-A committed:
[vcpkg/scripts/cmake] add missing $ and remove emtpy variable

Neumann-A committed:
[vcpkg/scripts/cmake] trigger full rebuild with a change to vcpkg_fixup_cmake_targets

Neumann-A commented:
hmm most regressions are failed due to download failures. The rest seems to be port build issues and not related to these changes.

JackBoosY commented:
Test result:
- sdl2:x64-uwp: failed on master branch.

Waiting for other pipeline results...

Neumann-A commented:
Hmm from the whole list only two ports seem to be relevant which are:
```
seal:x64-linux : needs further investigate
libfabric:x64-windows-static : needs further investigate
```
I was able to build both successfully using this branch.

JackBoosY commented:
@Neumann-A Yep, I also can't repro that in my machine.

JackBoosY commented:
OMG...more and more different regressions...

JackBoosY commented:
- [ ] osg:x64-linux
- [ ] signalrclient:x64-linux
- [ ] opencc:x64-linux
- [ ] seal:x64-linux
- [ ] xmsh:x64-osx
- [ ] rply:x64-osx
- [ ] seal:x64-osx
- [ ] signalrclient:x64-osx
- [ ] gmp:x64-osx
- [ ] rply:x64-uwp
- [ ] solid3:x64-uwp
- [ ] protobuf:x64-uwp
- [ ] signalrclient:x64-windows
- [ ] seal:x64-windows
- [ ] libaiff:x64-windows
- [ ] monkeys-audio:x64-windows
- [ ] rply:x64-windows
- [ ] pdal:x64-windows
- [ ] gmp:x64-windows
- [ ] dmlc:x64-windows
- [ ] icu:x64-windows
- [ ] signalrclient:x64-windows-static
- [ ] seal:x64-windows-static
- [ ] dcmtk:x64-windows-static
- [ ] icu:x64-windows-static
- [ ] rply:x64-windows-static
- [ ] pdal:x64-windows-static
- [ ] gmp:x64-windows-static
- [ ] zxing-cpp:x64-windows-static
- [ ] libfabric:x64-windows-static
- [ ] signalrclient:x86-windows
- [ ] seal:x86-windows
- [ ] monkeys-audio:x86-windows
- [ ] gsoap:x86-windows
- [ ] rply:x86-windows
- [ ] pdal:x86-windows
- [ ] gmp:x86-windows
- [ ] icu:x86-windows

JackBoosY commented:
/azp run

BillyONeal commented:
This PR always seems to be the first one that triggers msys losing it ¯\\\_(ツ)\_/¯

BillyONeal commented:
/azp run

BillyONeal commented:
Either the msys servers relied upon by some ports are broken causing CI failure, or something in this change causes CI to hang consistently.

BillyONeal commented:
This change depends on msys being fixed as in https://github.com/microsoft/vcpkg/pull/11443 or https://github.com/microsoft/vcpkg/pull/11471

Neumann-A commented:
@JackBoosY: None of those regressions look like they are caused by this PR. I assume some hidden unresolved dependencies in the ports which generate the errors in a full rebuild. (also see #11502 for comparison)

pdal: -> find_package issue. PostgreSQL::PostgreSQL in wrapper used without check if it actually exists.

JackBoosY commented:
@Neumann-A That's correct, I only want to know which port failed on the baseline.

Neumann-A commented:
@JackBoosY: Ah okay was just wondering due to the `requires:author-response` tag

JackBoosY commented:
/azp run

JackBoosY commented:
@Neumann-A pipelines are fixed, could you resolve the file conflict?

Thanks.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

# Conflicts:
#	scripts/cmake/vcpkg_build_cmake.cmake

Neumann-A commented:
@JackBoosY: done

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into build_variables

JackBoosY commented:
Can you merge the latest changes into this branch?

PhoebeHui commented:
/azp run

PhoebeHui committed:
Merge from master

PhoebeHui commented:
/azp run

PhoebeHui commented:
@Neumann-A, ocilib:x86-windows and ocilib:x64-windows failed with debug mode, I noticed it only build release mode before, the solution doesn't configure debug mode, could you help have a look?

       "D:\buildtrees\ocilib\x86-windows-dbg\v4.6.4-271d6c24cf\proj\dll\ocilib_dll_vs2019.sln" (Rebuild target) (1) ->
       (ValidateSolutionConfiguration target) -> 
         D:\buildtrees\ocilib\x86-windows-dbg\v4.6.4-271d6c24cf\proj\dll\ocilib_dll_vs2019.sln.metaproj : error MSB4126: The specified solution configuration "Debug|x86" is invalid. Please specify a valid solution configuration using the Configuration and Platform properties (e.g. MSBuild.exe Solution.sln /p:Configuration=Debug /p:Platform="Any CPU") or leave those properties blank to use the default solution configuration. [D:\buildtrees\ocilib\x86-windows-dbg\v4.6.4-271d6c24cf\proj\dll\ocilib_dll_vs2019.sln]

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

# Conflicts:
#	scripts/cmake/vcpkg_common_definitions.cmake

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

# Conflicts:
#	scripts/cmake/vcpkg_configure_cmake.cmake
#	scripts/cmake/vcpkg_configure_qmake.cmake

Neumann-A committed:
fix ocilib regression

JackBoosY commented:
`monkeys-audio` regression will be fixed in another PR.

strega-nil commented on scripts/cmake/vcpkg_common_definitions.cmake:
```diff
@@ -17,6 +17,13 @@
 ## VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX       import library suffix for target (same as CMAKE_IMPORT_LIBRARY_SUFFIX)
 ## VCPKG_FIND_LIBRARY_PREFIXES              target dependent prefixes used for find_library calls in portfiles
 ## VCPKG_FIND_LIBRARY_SUFFIXES              target dependent suffixes used for find_library calls in portfiles
+## VCPKG_BUILD_LIST                         List of VCPKG_BUILD_TYPE which the current port will build (uppercase)
+## VCPKG_BUILD_SHORT_NAME_<BUILDTYPE>       Short name of the buildtype (e.g. DEBUG=dbg; RELEASE=rel)
+## VCPKG_BUILD_CMAKE_TYPE_<BUILDTYPE>       CMAKE_BUILD_TYPE used for buildtype
+## VCPKG_BUILD_QMAKE_CONFIG_<BUILDTYPE>     Required QMAKE CONFIG flags for buildtype
+## VCPKG_PATH_SUFFIX_<BUILDTYPE>            Path suffix used for buildtype (e.g. /debug)
+## VCPKG_BUILD_TRIPLET_<BUILDTYPE>          Fullname of the buildtriplet e.g. ${TARGET_TRIPLET}-${VCPKG_BUILD_SHORT_NAME_<BUILDTYPE>}
+## VCPKG_BUILDTREE_TRIPLET_DIR_<BUILDTYPE>  Path to current buildtype buildtree (e.g. CURRENT_BUILDTREES_DIR/TRIPLET-rel )
```
In discussion, we have agreed on the following things:

1. We probably shouldn't define `VCPKG_BUILD_CMAKE_TYPE_*` and `VCPKG_BUILD_QMAKE_CONFIG_*`, since they're build-system specific.
2. We should remove `VCPKG_BUILDTREE_TRIPLET_DIR_*`, since that's not a concept we want to codify.
3. We need to rename these:
  * `VCPKG_BUILD_LIST` -> `VCPKG_BUILD_TYPES`
  * `VCPKG_BUILD_SHORT_NAME_*` -> `VCPKG_BUILD_TYPE_SHORT_NAME_*`
  * `VCPKG_BUILD_TRIPLET_*` -> `VCPKG_LOGNAME_STEM_*` or something similar

`VCPKG_PATH_SUFFIX_*` is good as it is.

Additionally, we want to lowercase the build types in these, so that `VCPKG_BUILD_TYPES` becomes `release;debug`,
and the different variables are `VCPKG_PATH_SUFFIX_release`, `VCPKG_PATH_SUFFIX_debug`, etc.

Neumann-A replied:
Please re-discuss the following or give arguments:
> Additionally, we want to lowercase the build types in these

uppercase is the way to go since cmake variables itself are uppercase. If I don't use uppercase build types I always have to use a `string(UPPER)` for `VCPKG_DETECTED_<SOMELANGUAGE>_FLAGS_<SOMEBUILDTYPE_UPPER> `. I don't see any good reason to stick to lowercase build types here. (I mean I can also introduce `VCPKG_BUILD_TYPE_UPPER_<config>` if you plan to stick to lowercase configs but it will get complicated to access flags: `MY_C_FLAGS_${VCPKG_BUILD_TYPE_UPPER_${config}}` instead of `MY_C_FLAGS_${config}`)

>We should remove VCPKG_BUILDTREE_TRIPLET_DIR_*, since that's not a concept we want to codify.

The path is used so often and even inside portfiles. I can rename it to `_VCPKG_INTERNAL_BUILDTREE_TRIPLET_DIR_`

> VCPKG_BUILD_TRIPLET_* -> VCPKG_LOGNAME_STEM_* or something similar

although often used for the logname it is not only used there. It is also used in messages and paths/folders. 

>We probably shouldn't define VCPKG_BUILD_CMAKE_TYPE_* and VCPKG_BUILD_QMAKE_CONFIG_*, since they're build-system specific.

TODO: I'll move them to the configure scripts.

strega-nil replied:
The thing is, `release` and `debug` aren't necessarily related to the CMake stuff, they're related to the vcpkg BUILD_TYPE stuff; even though the words are the same between VCPKG_BUILD_TYPE = release and CMAKE_BUILD_TYPE = Release, they're not necessarily the same thing, and we shouldn't use RELEASE just because CMake does for a related but different concept.

The big reason we want to do this is we wish to tie this to VCPKG_BUILD_TYPE. I think this would be a really unfortunate and weird feature if it weren't tied to VCPKG_BUILD_TYPE.

strega-nil commented on ports/3fd/portfile.cmake:
```diff
@@ -100,4 +100,4 @@ endif()
 file(INSTALL ${SOURCE_PATH}/Acknowledgements.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/3fd)
 
-vcpkg_copy_pdbs()
+vcpkg_copy_pdbs()
```
Should be removed from the diff.

strega-nil commented on ports/libqglviewer/portfile.cmake:
```diff
@@ -21,4 +21,4 @@ vcpkg_configure_qmake(
 
 file(INSTALL ${SOURCE_PATH}/QGLViewer DESTINATION ${CURRENT_PACKAGES_DIR}/include FILES_MATCHING PATTERN "*.h")
-file(INSTALL ${SOURCE_PATH}/LICENCE DESTINATION ${CURRENT_PACKAGES_DIR}/share/libqglviewer RENAME copyright)
+file(INSTALL ${SOURCE_PATH}/LICENCE DESTINATION ${CURRENT_PACKAGES_DIR}/share/libqglviewer RENAME copyright)
```
Should be removed from the diff.

strega-nil commented on ports/ocilib/portfile.cmake:
```diff
@@ -31,7 +31,7 @@ if(VCPKG_TARGET_IS_WINDOWS)
     # There is no debug configuration
     # As it is a C library, build the release configuration and copy its output to the debug folder
-    set(VCPKG_BUILD_TYPE release)
+    set(VCPKG_BUILD_LIST RELEASE)
```
Doesn't need to be done anymore.

strega-nil commented on ports/ocilib/CONTROL:
```diff
@@ -1,5 +1,6 @@
 Source: ocilib
 Version: 4.7.1
+Port-Version: 1
```
Thus, this stays the same as well

strega-nil commented on ports/qcustomplot/portfile.cmake:
```diff
@@ -35,4 +35,4 @@ file(INSTALL ${SOURCE_PATH}/qcustomplot.h
 
 # Handle copyright
-configure_file(${SOURCE_PATH}/GPL.txt ${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright COPYONLY)
+configure_file(${SOURCE_PATH}/GPL.txt ${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright COPYONLY)
```
Remove from the diff.

ras0219 commented on scripts/cmake/vcpkg_common_definitions.cmake:
```diff
@@ -127,6 +134,33 @@ set(CMAKE_IMPORT_LIBRARY_PREFIX "${VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX}")
+    set(VCPKG_BUILD_CMAKE_TYPE_${_buildname} "Release")
+    set(VCPKG_BUILD_QMAKE_CONFIG_${_buildname} "CONFIG+=release;CONFIG-=debug")
+    set(VCPKG_PATH_SUFFIX_${_buildname} "")
+    set(VCPKG_COPY_EXE_${_buildname} 1)
```
This should be set to `0` in the other branch.

strega-nil reviewed (changes requested):
Thanks @Neumann-A, just some minor last changes and a few removals/deletions.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

Neumann-A committed:
remove any unnecessary changes

Neumann-A committed:
rename VCPKG_BUILD_LIST to VCPKG_BUILD_TYPES
add VCPKG_COPY_EXE to 0 in the debug case.

Neumann-A committed:
lowercase buildname.
rename VCPKG_BUILD_SHORT_NAME_ to VCPKG_BUILD_TYPE_SHORT_NAME_
move VCPKG_BUILD_CMAKE_TYPE_ and VCPKG_BUILD_QMAKE_CONFIG_ into the the appropiate functions

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into build_variables

# Conflicts:
#	scripts/cmake/vcpkg_configure_cmake.cmake

Neumann-A commented on scripts/cmake/vcpkg_common_definitions.cmake:
```diff
@@ -17,6 +17,13 @@
 ## VCPKG_FIND_LIBRARY_SUFFIXES              target dependent suffixes used for find_library calls in portfiles
+## VCPKG_BUILD_TYPES                         List of VCPKG_BUILD_TYPE which the current port will build (uppercase)
+## VCPKG_BUILD_TYPE_SHORT_NAME_<BUILDTYPE>       Short name of the buildtype (e.g. DEBUG=dbg; RELEASE=rel)
+## VCPKG_BUILD_CMAKE_TYPE_<BUILDTYPE>       CMAKE_BUILD_TYPE used for buildtype
+## VCPKG_BUILD_QMAKE_CONFIG_<BUILDTYPE>     Required QMAKE CONFIG flags for buildtype
```
```suggestion
```

JackBoosY commented:
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.
